### PR TITLE
Add pre-commit-config and apply hooks

### PR DESCRIPTION
**Pull request by tsmbland**
This is based off the configuration in the [template](https://github.com/ImperialCollegeLondon/pip-tools-template/blob/main/.pre-commit-config.yaml), although I'm not using mypy, using markdownlint-fix instead of markdownlint, and a few more checks disabled.

Most of the changes have been automatically applied, although some manual fixes were required to make ruff happy.

All passes locally

### CODE REVIEW

File: .pre-commit-config.yaml
Comment (AdrianDAlessandro) [inline .pre-commit-config.yaml:1]: Can add some more to this. See the more up-to-date [python-template pre-commit config](https://github.com/ImperialCollegeLondon/python-template/blob/main/%7B%7B%20cookiecutter.project_slug%20%7D%7D/.pre-commit-config.yaml)

**Review by AdrianDAlessandro — CHANGES_REQUESTED** [review]
I've made a suggestion to add some more pre-commit tools to the config.

Also, since #57 was merged, this now has merge conflicts. Can you resolve those?

**Review by Thomas-Rowlands — APPROVED** [review]

File: .pre-commit-config.yaml
Comment (tsmbland) [reply .pre-commit-config.yaml:1]: Ahhh I didn't realise there was a new template. Good to know! Will do!

**Comment by Thomas-Rowlands** [comment]
Conflicts merged, most of the linting changes made it in except any changes that would revert the Pathlib PR.

**Comment by tsmbland** [comment]
Thanks @Thomas-Rowlands! I'm going through @AdrianDAlessandro's suggestion now to add a few more tools

**Comment by alexdewar** [comment]
Can we just fix the whitespace in `src/IAO_dicts`? It's just static data in there, so it shouldn't be affected by whitespace....

File: tests/data/PMC8885717.html
Comment (alexdewar) [inline tests/data/PMC8885717.html:1]: I'm being a bit pedantic here, but might we want to exclude this from all the linters including whitespace-related ones?

It's just that it is "real" data in the sense that this is (presumably) the form it was in when it was downloaded. I doubt it matters much though.

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -0,0 +1,37 @@
+repos:
+  - repo: https://github.com/pre-commit/pre-commit-hooks
+    rev: v4.6.0
+    hooks:
+      - id: check-merge-conflict
+      - id: debug-statements
+      - id: trailing-whitespace
+        exclude: ^src/IAO_dicts/
```
Comment (alexdewar) [inline .pre-commit-config.yaml:8]: Did we figure out why we to exclude this folder? If trimming the whitespace breaks the parser, it seems like it's the parser's problem. (Admittedly, fixing that is out of scope for this PR!)

**Review by alexdewar — APPROVED** [review]
I've had a skim and it seems fine to me, except for a couple of really minor questions I have.

File: src/IAO_dicts/IAO_FINAL_MAPPING.txt [hunk 2]
```diff
@@ -114,7 +114,7 @@ experimental design	methods section / materials section
 experimental methods	methods section
 experimental procedures	methods section
 experimental section	methods section
-extented data	supplementary material section
+extended data	supplementary material section
```
[the remark this answers is not in the document]
Comment (alexdewar) [reply src/IAO_dicts/IAO_FINAL_MAPPING.txt:117]: I'm assuming not. **But** there's a possibility that the misspelling is deliberate (i.e. there are actual papers out there containing the phrase "extented data").

@jmp111 @tb143 @Thomas-Rowlands Can you advise?

File: .pre-commit-config.yaml (hunk 1 above)
Comment (AdrianDAlessandro) [reply .pre-commit-config.yaml:8]: Nah, it's chill

Comment (alexdewar) [reply .pre-commit-config.yaml:8]: Word

File: tests/data/PMC8885717.html
Comment (AdrianDAlessandro) [reply tests/data/PMC8885717.html:1]: We will merge this PR and undo all the changes to the test data separately

**Review by AdrianDAlessandro — APPROVED** [review]
